<launch>
    
    <node pkg="camera1394stereo" type="camera1394stereo_node" name="camera1394stereo_node" output="screen" >
	    <param name="video_mode" value="format7_mode3" />
	    <param name="format7_color_coding" value="raw16" />
	    <param name="bayer_pattern" value="bggr" />
	    <param name="bayer_method" value="" />
	    <param name="stereo_method" value="Interlaced" />
	    <param name="camera_info_url_left" value="" />
	    <param name="camera_info_url_right" value="" />
  	</node>
    	
  	<arg name="pi/2" value="1.5707963267948966" />
    <arg name="optical_rotate" value="0 0 0 -$(arg pi/2) 0 -$(arg pi/2)" />
  	<node pkg="tf" type="static_transform_publisher" name="camera_base_link"
        args="$(arg optical_rotate) base_link stereo_camera 100" />  
  	
  	<!-- Run the ROS package stereo_image_proc -->
	<group ns="/stereo_camera" >
		<!-- Throttle camera images to 10 Hz -->
		<node pkg="nodelet" type="nodelet" name="stereo_throttle" args="standalone rtabmap/stereo_throttle">
			<remap from="left/image"       to="left/image_raw"/>
		    <remap from="right/image"      to="right/image_raw"/>
		    <remap from="left/camera_info" to="left/camera_info"/>
		    <remap from="right/camera_info" to="right/camera_info"/>
		    
		    <param name="queue_size" type="int" value="10"/>
		    <param name="rate" type="double" value="10"/>
		</node>
	
		<node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc">
		<remap from="left/image_raw" to="left/image_raw_throttle"/>
		<remap from="left/camera_info" to="left/camera_info_throttle"/>
		<remap from="right/image_raw" to="right/image_raw_throttle"/>
		<remap from="right/camera_info" to="right/camera_info_throttle"/>
		</node>
	</group>

	<!-- Odometry: Choose between viso2_ros, fovis_ros and homemade approach (default) -->	
	<node pkg="viso2_ros" type="stereo_odometer" name="stereo_odometer" output="screen">
		<remap from="stereo" to="/stereo_camera"/>
		<remap from="image" to="image_rect"/>
		<param name="base_link_frame_id" value="base_link"/>
		<param name="odom_frame_id" value="/odom"/>
		<param name="ref_frame_change_method" value="1"/>
		<remap from="left/camera_info" to="left/camera_info_throttle" />
	    <remap from="right/camera_info" to="right/camera_info_throttle" />
	</node>
	
	<!--
	<node pkg="fovis_ros" type="fovis_stereo_odometer" name="stereo_odometer" >
	    <remap from="/stereo/left/image" to="/stereo_camera/left/image_rect" />
	    <remap from="/stereo/right/image" to="/stereo_camera/right/image_rect" />
	    <remap from="/stereo/left/camera_info" to="/stereo_camera/left/camera_info_throttle" />
	    <remap from="/stereo/right/camera_info" to="/stereo_camera/right/camera_info_throttle" />
	    <remap from="odometry" to="stereo_odometer/odometry" />
    </node>
    -->	
    <!--
    <node pkg="rtabmap" type="stereo_odometry" name="stereo_odometer" output="screen">
  	  	<remap from="left/image_rect"       to="/stereo_camera/left/image_rect"/>
  	  	<remap from="right/image_rect"      to="/stereo_camera/right/image_rect"/>
  	  	<remap from="left/camera_info"      to="/stereo_camera/left/camera_info_throttle"/>
  	  	<remap from="right/camera_info"     to="/stereo_camera/right/camera_info_throttle"/>
	  	<remap from="odom"                  to="/stereo_odometer/odometry"/>
    
  	  	<param name="frame_id" type="string" value="base_link"/>
  	  	<param name="approx_sync" type="bool" value="false"/>
  	
	  	<param name="Odom/Strategy" type="string" value="1"/>            
		<param name="Odom/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
		<param name="Odom/MaxDepth" type="string" value="10"/>
  	</node>
  	-->
    
  <group ns="rtabmap">	
    <!-- Visual SLAM (robot side) -->
    <!-- args: "delete_db_on_start" and "udebug" -->
    <node name="rtabmap" pkg="rtabmap" type="rtabmap" output="screen" args="--delete_db_on_start">
      <param name="subscribe_depth" type="bool" value="false"/>
      <param name="subscribe_stereo" type="bool" value="true"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
	
      <remap from="left/image_rect" to="/stereo_camera/left/image_rect_color"/>
      <remap from="right/image_rect" to="/stereo_camera/right/image_rect"/>
      <remap from="left/camera_info" to="/stereo_camera/left/camera_info_throttle"/>
      <remap from="right/camera_info" to="/stereo_camera/right/camera_info_throttle"/>
      
      <remap from="odom" to="/stereo_odometer/odometry"/>
      
      <param name="frame_id" type="string" value="base_link"/>
      <param name="queue_size" type="int" value="30"/>
      
      <param name="Rtabmap/TimeThr" type="string" value="700"/>
      <param name="Rtabmap/DetectionRate" type="string" value="1"/>
      <param name="Kp/DetectorStrategy" type="string" value="0"/>
      <param name="Kp/WordsPerImage" type="string" value="200"/>
      <param name="Kp/RoiRatios" type="string" value="0.03 0.03 0.04 0.04"/>
      <param name="NN/NNStrategy" type="string" value="1"/>
      
      <param name="LccBow/MinInliers" type="string" value="10"/>
      <param name="LccBow/InlierDistance" type="string" value="0.02"/>
      <param name="LccBow/MaxDepth" type="string" value="3"/>
      
      <param name="LccReextract/FeatureType" type="string" value="4"/>
	  <param name="LccReextract/Activated" type="string" value="true"/>
	  <param name="LccReextract/MaxDepth" type="string" value="10"/>
	  <param name="LccReextract/NNDR" type="string" value="0.8"/>
	  <param name="LccReextract/NNType" type="string" value="3"/> 		  	
    </node>
  
    <!-- Visualisation (client side) -->
    <!--
    <node pkg="rtabmap" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_depth" type="bool" value="false"/>
      <param name="subscribe_laserScan" type="bool" value="false"/>
      <param name="queue_size" type="int" value="30"/>
    
      <remap from="rgb/image" to="/stereo_camera/left/image_rect_color"/>
      <remap from="rgb/camera_info" to="/stereo_camera/left/camera_info_throttle"/>
      <remap from="depth/image" to="/odom_depth"/>
      
      <remap from="odom" to="/stereo_odometer/odometry"/>
    </node>
    -->
  </group>
  
  <!-- RVIZ -->
  
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap)/launch/config/rgbd.rviz"/>
  
</launch>